### PR TITLE
[velux] first version of the OH1 binding

### PR DESCRIPTION
**Pull request by gs4711**
This binding integrates the Velux devices with help of a special gateway, the Velux Bridge KLF200. The Velux Binding interacts via the Velux Bridge with the Velux devices like controlling window openers, shutters and others.

In addition to the OH2 binding it supports a virtual rollershutter consisting of different scene definitions which lead to the desired rollershutter level.

### CODE REVIEW

**Comment by 9037568** [comment]
Did you intend this to be a WIP, @gs4711 ?
Or do you consider it ready for review?

**Comment by gs4711** [comment]
Hello Chris,
this binding works fine for OH1 and it should have been a push for review.
But it seems to be that Jenkins does not feel comfortable due to some issues of another binding.
Best regards, Guenther

**Comment by 9037568** [comment]
Looks like a transient failure to me. Let's retry it.

**Comment by 9037568** [comment]
Ok, guess not. Looks like the problem is in your pom.xml.
Need to remove the `<relativePath>` element.

**Comment by gs4711** [comment]
@9037568 Thanks. Good point.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/VeluxRSBindingConfig.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+ * as declared in the binding configuration (possibly adapted by preprocessing).</li>
+ * </ul>
+ *
+ * @author Guenther Schreiner
+ * @since 1.13.0
+ */
+
+public class VeluxRSBindingConfig extends VeluxBindingConfig {
+
+    static final Logger logger = LoggerFactory.getLogger(VeluxRSBindingConfig.class);
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/VeluxRSBindingConfig.java]: "private final", not "static final"

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/config/VeluxBridgeConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * <li>{@link VeluxBridgeConfiguration#bridgePassword bridgePassword} brigde password,</li>
+ * <li>{@link VeluxBridgeConfiguration#timeoutMsecs timeoutMsecs} communication timeout in milliseconds,</li>
+ * <li>{@link VeluxBridgeConfiguration#retries retries} number of retries (with exponential backoff algo),</li>
+ * <li>{@link VeluxBridgeConfiguration#refresh refresh} refresh interval for retrieval of bridge information.</li>
+ * </ul>
+ * <p>
+ *
+ * @author Guenther Schreiner - Initial contribution
+ */
+// FIXME: for OpenHAB2 the following line is acceptable:
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/config/VeluxBridgeConfiguration.java]: Delete this comment block

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/config/VeluxBridgeConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    public int bridgeTCPPort;
+    public String bridgePassword;
+    public int timeoutMsecs;
+    public int retries;
+    public long refresh;
+
+    /**
+     * Default values - should not be modified
+     */
+    public VeluxBridgeConfiguration() {
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/config/VeluxBridgeConfiguration.java]: Move these values into the declarations and then drop this constructor.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/config/VeluxThingConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * It contains the factory default values as well.
+ * <p>
+ *
+ * @author Guenther Schreiner - Initial contribution
+ */
+public class VeluxThingConfiguration {
+
+    public static final String THING_SCENE_NAME = VeluxBindingProperties.PROPERTY_SCENE_NAME;
+
+    // public String sceneName;
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/config/VeluxThingConfiguration.java]: Deleted this commented code block

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxExistingProducts.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    // Type definitions
+
+    // private ConcurrentHashMap<String, VeluxProduct> existingProductsByName;
+    private ConcurrentHashMap<String, VeluxProduct> existingProductsByUniqueIndex;
+    private int memberCount;
+
+    public VeluxExistingProducts() {
+        logger.trace("VeluxExistingProducts() initializing.");
+        // existingProductsByName = new ConcurrentHashMap<String, VeluxProduct>();
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxExistingProducts.java]: Delete this commented line.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProduct.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+    public VeluxProduct(BCgetProducts.BCproduct productDescr) {
+        this.name = new ProductName(productDescr.getName());
+        this.bridgeProductIndex = new ProductBridgeIndex(productDescr.getId());
+        String productTypeId = productDescr.getCategory().replaceAll("\\s", "_").toUpperCase();
+
+        try {
+            this.typeId = ProductTypeId.valueOf(productTypeId);
+        } catch (IllegalArgumentException iae) {
+            logger.error(
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProduct.java]: This should be a WARN, not an ERROR.
Also, drop the ALL CAPS text.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProduct.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+                    "PLEASE REPORT THIS TO MAINTAINER: VeluxProduct({},{}) has found an unregistered ProductTypeId.",
+                    productTypeId, productDescr.getTypeId());
+        }
+        if (this.getTypeId().getTypeId() != productDescr.getTypeId()) {
+            logger.error(
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProduct.java]: This should be a WARN, not an ERROR.
Also, drop the ALL CAPS text.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductReference.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+ * @see VeluxProduct
+ *
+ * @author Guenther Schreiner - initial contribution.
+ */
+public class VeluxProductReference {
+    private final Logger logger = LoggerFactory.getLogger(VeluxProductReference.class);
+
+    // Type definitions
+
+    // imported public static class ProductName
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductReference.java]: Drop these commented code lines

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductReference.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    private ProductName name;
+    private ProductTypeId typeId;
+
+    // Constructor
+
+    public VeluxProductReference(BCproductState productState) {
+        this.name = new ProductName(productState.getName());
+        this.typeId = ProductTypeId.get(productState.getTypeId());
+        if (this.typeId == null) {
+            logger.error(
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductReference.java]: This should be a WARN, not an ERROR.
Also, drop the ALL CAPS text.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductState.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+ * <B>Velux</B> product status representation.
+ * <P>
+ * Combined set of information which describes a current state of a single Velux product.
+ *
+ * @see VeluxProduct
+ *
+ * @author Guenther Schreiner - initial contribution.
+ */
+public class VeluxProductState {
+    // private static final Logger LOGGER = LoggerFactory.getLogger(VeluxProductState.class);
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductState.java]: Delete or uncomment this line.
If uncommenting, drop the "static" keyword.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductState.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+    // Type definitions
+
+    public class ProductState {
+        //
+        // Meaningful values (TBD as enum):
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductState.java]: Delete these commented code lines.

**Comment by gs4711** [comment]
Thanks a lot for the very thoughtful comments and suggestions.

Anything hopefully fixed well, but one point: 

- the removal of the dependancy to <code>com.google.code.gson</code> within <code>pom.xml</code> by including it into <code>META-INF/MANIFEST.MF</code> fails within my Eclipse as well as using maven.

**Comment by 9037568** [comment]
True, if you _just_ remove the dependency, your build will fail.
You have to add the library jar to the `lib` folder and put a reference in the MANIFEST and .classpath.
You can look at how it's done in the [mystromecopower binding](https://github.com/openhab/openhab1-addons/tree/master/bundles/binding/org.openhab.binding.mystromecopower), for example:

* [https://github.com/openhab/openhab1-addons/tree/master/bundles/binding/org.openhab.binding.mystromecopower/lib](https://github.com/openhab/openhab1-addons/tree/master/bundles/binding/org.openhab.binding.mystromecopower/lib)
* [https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.mystromecopower/META-INF/MANIFEST.MF](https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.mystromecopower/META-INF/MANIFEST.MF)
* [https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.mystromecopower/.classpath](https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.mystromecopower/.classpath)

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 12]
```diff
@@ -1,90 +1,80 @@
...
 
 ## Binding Configuration
 
-The binding can be configured by parameters in the global configuration file `/etc/openhab/configurations/openhab.cfg`.
+The binding can be configured by parameters in the global configuration file `openhab.cfg`.
 
 | Property       | Default                | Required | Description                                           |
 |----------------|------------------------|:--------:|-------------------------------------------------------|
-| bridgeIPAddress| 127.0.0.1              |   Yes    | Name of address for accessing the Velux Bridge.       |
+| bridgeIPAddress| 127.0.0.1              |   Yes    | Hostname or address for accessing the Velux Bridge.   |
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: Required still needs to be a `No` here.

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 13]
```diff
@@ -1,90 +1,80 @@
...
-
-| Firmware revision | Release date | Description                                                             |
-|:-----------------:|:------------:|-------------------------------------------------------------------------|
-| 0.1.1.0.41.0      | 2016-06-01   | Default factory shipping revision.                                      |
-| 0.1.1.0.42.0      | 2017-07-01   | N/A                                                                     |
-| 0.1.1.0.44.0      | 2017-12-14   | N/A                                                                     |
-
 ## Discovery
 
 Unfortunatelly there is no way to discover the Velux bridge within the local network. Beware that all Velux scenes have to be added to the local Velux Bridge configuration as described in the Velux setup procedure.
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: "Unfortunately"

Also, "Beware" should probably say "Be aware"

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 14]
```diff
@@ -94,27 +84,16 @@ As the bridge does not support a real rollershutter interaction, this binding pr
...
- * Copyright (c) 2010-2018 by the respective copyright holders.
- *
- * All rights reserved. This program and the accompanying materials
- * are made available under the terms of the Eclipse Public License v1.0
- * which accompanies this distribution, and is available at
- * http://www.eclipse.org/legal/epl-v10.html
- */
-
 /**
  * OpenHAB item defintion for velux binding:
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: Are these Javadoc style comments supported in a .items file? If not, they should be converted to the standard // style.

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 15]
```diff
@@ -136,13 +115,13 @@ Switch  V_DG_OPEN   "Velux DG open"             (gV)    { velux="thing=scene;cha
 Switch  V_DG_SUNNY  "Velux DG sunny"            (gV)    { velux="thing=scene;channel=action#V_DG_Shutters_090" }
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: These binding definitions don't match the one given in the "Item Configuration" section above, which says:

````
{ velux="thing=<Mastertype>channel=<Subtype>" }
````

Please review the above definition and correct it.

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 16]
```diff
@@ -94,27 +84,16 @@ As the bridge does not support a real rollershutter interaction, this binding pr
 ### Items velux.items
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: This line should say only "Items".  "velux.items" belongs above the code fence that starts the set of examples.

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 17]
```diff
@@ -152,10 +131,6 @@ Rollershutter V_DG_M_S  "Velux DG Rolladen Mitte [%d]"  { velux="thing=bridge;ch
 0,100,V_DG_Shutter_Mitte_100"}
 Rollershutter V_DG_M_W  "Velux DG Window Mitte [%d]"    { velux="thing=bridge;channel=shutter#0,V_DG_Window_Mitte_000,10,V_DG_Window_Mitte_010,
 100,V_DG_Window_Mitte_100"}
-
-//
-// end-of-items/velux.items
-//
 ```
 
 ### Sitemap velux.sitemap
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: This line should say just "Sitemap". "velux.sitemap" belongs above the code fence that starts the set of examples.

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 18]
```diff
@@ -164,28 +139,24 @@ Rollershutter V_DG_M_W  "Velux DG Window Mitte [%d]"    { velux="thing=bridge;ch
...
     Frame label="Velux Bridge" {
         Text    item=V_CHECK
         Text    item=V_STATUS
         Text    item=V_FIRMWARE
     }
-    
 }
 ```
 
 ### Rule velux.rules
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: This line should say just "Rules". "velux.rules" belongs above the code fence that starts the set of examples.

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 19]
```diff
@@ -203,34 +174,31 @@ rule "PushButton of group gV"
...
-    </logger>
+[INFO ] [.b.velux.internal.VeluxBinding] - Velux refresh interval is 3600000 seconds.
+[INFO ] [.service.AbstractActiveService] - velux Refresh Service has been started
+[INFO ] [.b.velux.internal.VeluxBinding] - veluxConfig[bridgeIPAddress=192.168.41.1,bridgeTCPPort=80,bridgePassword=********,timeoutMsecs=2000,retries=6]
 ```
 
+## Supported/Tested Firmware Revisions
 
-During startup of normal operations, there should be only some few messages within the logfile, like:
+The Velux Bridge in API version One (firmware version 0.1.1.*) allows to activate a set of predefined actions, so called scenes. Therefore besi
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: Join this line with the line below.

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 20]
```diff
@@ -203,34 +174,31 @@ rule "PushButton of group gV"
...
+de the bridge, only one main thing exists, the scene element. Unfortunatelly even the current firmware version 0.1.1.0.44.0 does not include en
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: Join this line with the line below.

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 21]
```diff
@@ -203,34 +174,31 @@ rule "PushButton of group gV"
...
+|:-----------------:|:------------:|-------------------------------------------------------------------------|
+| 0.1.1.0.41.0      | 2016-06-01   | Default factory shipping revision.                                      |
+| 0.1.1.0.42.0      | 2017-07-01   | N/A                                                                     |
+| 0.1.1.0.44.0      | 2017-12-14   | N/A                                                                     |
 
-## Famous last words
+## Unknown Velux devices
 
-All known <B>Velux<B> devices can be handled by this binding. In fact, there might be some new one which will be reported within the logfiles. Therefore, if you recognize an error message like:
+All known <B>Velux<B> devices can be handled by this binding. However, there might be some new ones which will be reported within the logfiles.Therefore, error messages like the one below should be reported to the maintainers so that the new Velux device type can be incorporated."
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/README.md]: You have two opening `<B>` elements here.  The second one should be a closing `</B>`.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/VeluxBindingConstants.java [hunk 22]
```diff
@@ -14,13 +14,13 @@
...
  *
- * For an in-depth view of the available , take a look onto {@link VeluxItemType}.
+ * For an in-depth view of the available Item type with description of parameters, take a look onto {@link VeluxItemType}.
  *
  * @author Guenther Schreiner - Initial contribution
  */
 public class VeluxBindingConstants {
 
-    /** Basis bindung identification */
+    /** Basic bindung identification */
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/VeluxBindingConstants.java]: "bindung" should be "binding"

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/VeluxItemType.java [hunk 23]
```diff
@@ -112,7 +106,8 @@
     private boolean itemIsWritable;
     private boolean itemIsExecutable;
 
-    private static Logger logger = LoggerFactory.getLogger(VeluxItemType.class);
+    private final Logger logger = LoggerFactory.getLogger(VeluxItemType.class);
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/VeluxItemType.java]: The `logger_` variable should be removed, and all references to it converted back to references to `logger`.

I didn't notice before that this class contains a `static` method.  If that method needs to remain static, then it's ok to leave the logger as static also.

File: bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductState.java [hunk 24]
```diff
@@ -1,5 +1,4 @@
-/**
- * Copyright (c) 2010-2018 by the respective copyright holders.
+/** * Copyright (c) 2010-2018 by the respective copyright holders.
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/things/VeluxProductState.java]: Not sure what happened to this copyright header, but please revise it to match the other ones.

**Comment by gs4711** [comment]
Thank you for the quick feedback - and the good pointer to the other binding relating to including gson libraries.

**Comment by gs4711** [comment]
This was new to me. I'm thinking it looks good now.

**Comment by gs4711** [comment]
@9037568 👍 Many thanks for the ultra-quick responses and the very constructive feedback!

**Comment by 9037568** [comment]
Thanks, @gs4711 !

**Pull request merged**